### PR TITLE
Fix device when loading transformers model

### LINKED ISSUE #12871: [BUG] The Device parameter has no effect when loading an existing model from the registry with MLflow > 2.3.1
### Issues Policy acknowledgement

- [X] I have read and agree to submit bug reports in accordance with the [issues policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md)

### Where did you encounter this bug?

Databricks

### Willingness to contribute

No. I cannot contribute a bug fix at this time.

### MLflow version

> 2.3.1  (tested with 2.9.2 and 2.13.1)

### System information

- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
- **Python version**:
- **yarn version, if running the dev UI**:

### Describe the problem

I'm updating some older code that runs fine on Databricks runtime 13.1 (MLflow 2.3.1), but not on newer runtimes such as 14.3 (MLflow 2.9.2) or 15.4 Beta (MLflow 2.13.1).  I've tracked down the issue to loading the transformer model.  The model was trained on a GPU, but it needs to be loaded and run on a CPU endpoint. I've provided a minimal snippet that demonstrates the issue.

In DBR 13.1, the display statement returns `device(type='cpu')`.  In DBR 14.3, it returns `device(type='cuda', index=0)`

This is preventing me from wrapping the model and serving it on a CPU-only Databricks model serving endpoint.

I have tried `pipeline_cpu.model = pipeline_cpu.model.cpu()`, but this causes other issues with the input image data not being on the CPU.

### Tracking information

_No response_

### Code to reproduce issue

```python
import mlflow
import torch

model_uri = "models:/<path-to-model>"

pipeline_cpu = mlflow.transformers.load_model(
  model_uri, 
  return_type="pipeline", 
  device=torch.device("cpu").index)
display(pipeline_cpu.device)
```

### Stack trace

```
/root/.ipykernel/39407/command-1643111387547108-131983805:7: FutureWarning: The 'transformers' MLflow Models integration is known to be compatible with the following package version ranges: ``4.25.1`` -  ``4.35.2``. MLflow Models integrations with transformers may not succeed when used with package versions outside of this range.
  pipeline_cpu = mlflow.transformers.load_model(
/databricks/python/lib/python3.10/site-packages/mlflow/store/artifact/utils/models.py:32: FutureWarning: ``mlflow.tracking.client.MlflowClient.get_latest_versions`` is deprecated since 2.9.0. Model registry stages will be removed in a future major release. To learn more about the deprecation of model registry stages, see our migration guide here: https://mlflow.org/docs/2.9.2/model-registry.html#migrating-from-stages
  latest = client.get_latest_versions(name, None if stage is None else [stage])
 
/databricks/python/lib/python3.10/site-packages/mlflow/store/artifact/utils/models.py:32: FutureWarning: ``mlflow.tracking.client.MlflowClient.get_latest_versions`` is deprecated since 2.9.0. Model registry stages will be removed in a future major release. To learn more about the deprecation of model registry stages, see our migration guide here: https://mlflow.org/docs/2.9.2/model-registry.html#migrating-from-stages
  latest = client.get_latest_versions(name, None if stage is None else [stage])
2024/07/27 19:46:04 INFO mlflow.transformers: 'models:/dbdemos_pcb_classification/Production' resolved as 'dbfs://databricks/databricks/mlflow-registry/9e0fe40de9e0416b8118416ced05ad2e/models/model'


2024-07-27 19:46:06.278234: I tensorflow/core/util/port.cc:111] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2024-07-27 19:46:06.324822: E tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc:9342] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2024-07-27 19:46:06.324859: E tensorflow/compiler/xla/stream_executor/cuda/cuda_fft.cc:609] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2024-07-27 19:46:06.324887: E tensorflow/compiler/xla/stream_executor/cuda/cuda_blas.cc:1518] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2024-07-27 19:46:06.333386: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX512F AVX512_VNNI FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
/databricks/python/lib/python3.10/site-packages/transformers/models/vit/feature_extraction_vit.py:28: FutureWarning: The class ViTFeatureExtractor is deprecated and will be removed in version 5 of Transformers. Please use ViTImageProcessor instead.
  warnings.warn(

device(type='cuda', index=0)
```

### Other info / logs

_No response_

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [X] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [X] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by serena-ruan**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/serena-ruan/mlflow/pull/12977?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/12977/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 12977
```

</p>
</details>

### Related Issues/PRs

Resolve #12871

### What changes are proposed in this pull request?
Set default value of MLFLOW_HUGGINGFACE_USE_DEVICE_MAP to False. Emit warning if device is set but MLFLOW_HUGGINGFACE_USE_DEVICE_MAP is set to True.

### How is this PR tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [x] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

### CODE REVIEW

File: mlflow/transformers/__init__.py [hunk 1]
```diff
@@ -1183,8 +1175,22 @@ def _load_model(path: str, flavor_config, return_type: str, device=None, **kwarg
         conf["device_map"] = device_map_strategy
         accelerate_model_conf["device_map"] = device_map_strategy
         # Cannot use device with device_map
+        if device is not None:
+            _logger.warning(
+                "MLFLOW_HUGGINGFACE_USE_DEVICE_MAP is set to True, device "
+                f"is incompatible with device_map, ignoring device param `{device}` passed."
```
Comment (B-Step62) [inline mlflow/transformers/__init__.py:1181]: Shall we raise here? Now that L1170-1177 code block is moved to after this check, then `device` parameter is always opt-in.

File: mlflow/transformers/__init__.py [hunk 2]
```diff
@@ -1183,8 +1175,22 @@ def _load_model(path: str, flavor_config, return_type: str, device=None, **kwarg
...
         device = None
 
+    if device is None:
+        if MLFLOW_DEFAULT_PREDICTION_DEVICE.get():
+            try:
+                device = int(MLFLOW_DEFAULT_PREDICTION_DEVICE.get())
+            except ValueError:
+                device = _TRANSFORMERS_DEFAULT_CPU_DEVICE_ID
+        elif is_gpu_available():
+            device = _TRANSFORMERS_DEFAULT_GPU_DEVICE_ID
```
Comment (B-Step62) [inline mlflow/transformers/__init__.py:1192]: Doesn't this change the behavior when `MLFLOW_HUGGINGFACE_USE_DEVICE_MAP` is set to `True`?Previously, `device` is set to `None`, however, now it is overwritten here and one of these default values will be passed as `device` key of the loading config. Is it intended?

**Review by B-Step62 — COMMENTED** [review]
Can we add tests with mock pipeline to validate the correct parameters are passed for instantiation? It seems current tests do not capture change like [this](https://github.com/mlflow/mlflow/pull/12977#discussion_r1722834193).

File: mlflow/transformers/__init__.py (hunk 2 above)
Comment (serena-ruan) [reply mlflow/transformers/__init__.py:1192]: Oh right! Thanks for capturing this! Let me add a test

File: mlflow/transformers/__init__.py (hunk 1 above)
Comment (serena-ruan) [reply mlflow/transformers/__init__.py:1181]: Why do we need to raise? Any concerns about warning and ignoring?

File: tests/transformers/test_transformers_model_export.py [hunk 3]
```diff
@@ -3633,3 +3633,14 @@ def test_peft_pipeline_copy_metadata_in_databricks(mock_is_in_databricks, peft_p
     else:
         assert not os.path.exists(metadata_path)
     mock_is_in_databricks.assert_called_once()
+
+
+@pytest.mark.parametrize("device", ["cpu", "cuda", 0, -1, None])
```
Comment (harupy) [inline tests/transformers/test_transformers_model_export.py:3638]: Just out of curiosity, is `-1` a valid value for `device`?

File: mlflow/transformers/__init__.py (hunk 1 above)
Comment (B-Step62) [reply mlflow/transformers/__init__.py:1181]: Sure, here are the reasons I prefer exception:
1. It is incorrect usage of the API.
2. When users specify device, it means they want to load the model onto that specific devices. Ignoring it results in undesired outcome e.g. slow performance or OOM.
3. The priority between device map and device is not obvious to users. It is extra complexity we adds on top of transformer.

Comment (BenWilson2) [reply mlflow/transformers/__init__.py:1181]: +1 to raising. With a 1+ hour start-up time for many of these models in model serving, having a fast-fail for incompatible config is preferred. Can we make the exception message very clear about the incompatibility and how to fix it?

File: tests/transformers/test_transformers_model_export.py (hunk 3 above)
Comment (serena-ruan) [reply tests/transformers/test_transformers_model_export.py:3638]: Yes https://github.com/mlflow/mlflow/blob/3267c372f2dbb98d5f5870f3d7262f5b7591ecd0/mlflow/transformers/__init__.py#L149-L152

**Review by B-Step62 — APPROVED** [review]
LGTM!

File: mlflow/transformers/__init__.py [hunk 4]
```diff
@@ -1183,7 +1175,23 @@ def _load_model(path: str, flavor_config, return_type: str, device=None, **kwarg
...
+                "to False to specify a particular device ID, or pass None for the `device` "
+                "argument to use device_map."
+            )
         device = None
+    elif device is None:
+        if MLFLOW_DEFAULT_PREDICTION_DEVICE.get():
+            try:
+                device = int(MLFLOW_DEFAULT_PREDICTION_DEVICE.get())
+            except ValueError:
+                device = _TRANSFORMERS_DEFAULT_CPU_DEVICE_ID
```
Comment (harupy) [inline mlflow/transformers/__init__.py:1192]:
```suggestion
        if device_value := MLFLOW_DEFAULT_PREDICTION_DEVICE.get():
            try:
                device = int(device_value)
            except ValueError:
                _logger.warning(f"Invalid value for {MLFLOW_DEFAULT_PREDICTION_DEVICE}: {device_value!r}, setting device to {_TRANSFORMERS_DEFAULT_CPU_DEVICE_ID}")
                device = _TRANSFORMERS_DEFAULT_CPU_DEVICE_ID
```

can we insert a warning?

**Review by harupy — APPROVED** [review]
LGTM!

File: mlflow/transformers/__init__.py (hunk 4 above)
Comment (serena-ruan) [reply mlflow/transformers/__init__.py:1192]: Oops missed this, follow-up: https://github.com/mlflow/mlflow/pull/12991

**Pull request merged**